### PR TITLE
Fix semaphore issues updating CollectionView on iOS

### LINKED ISSUE #11853: [Bug][iOS] Concurrent issue leading to crash in SemaphoreSlim.Release in ObservableItemsSource
### Description
Calls to `ObservableItemsSource.BatchUpdate` should be synchronized to prevent a crash in _batchUpdating.Release(), because this last one can be called 2 times in a row and will fail the second time.

https://github.com/xamarin/Xamarin.Forms/blob/main/Xamarin.Forms.Platform.iOS/CollectionView/ObservableItemsSource.cs

### Steps to Reproduce
I am able to reproduce this issue easily in my customer's app. 
I've already checked that all calls are done on the main thread.

It happens nearly everytime when updating the bound ObservableCollection 3 times in a row: 
1st deleting some items, 2nd inserting some new items, 3rd moving some existing items (it's an algorithm that compute the changes between an ObservableCollection and a "new" list, and apply the 3 above actions so after the ObservableCollection becomes equal to the list).

I've already checked that all 3 calls are done on the main thread.

### Expected Behavior
don't crash

### Actual Behavior
crash when releasing the SemaphoreSlim because its CurrentCount is already 1.

Stack trace:
```
SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.

  at System.Threading.SemaphoreSlim.Release (System.Int32 releaseCount) [0x0004c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/SemaphoreSlim.cs:795 
  at System.Threading.SemaphoreSlim.Release () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/SemaphoreSlim.cs:760 
  at Xamarin.Forms.Platform.iOS.ObservableItemsSource.Reload () [0x00098] in D:\a\1\s\Xamarin.Forms.Platform.iOS\CollectionView\ObservableItemsSource.cs:146 
  at Xamarin.Forms.Platform.iOS.ObservableItemsSource.Add (System.Collections.Specialized.NotifyCollectionChangedEventArgs args) [0x00037] in D:\a\1\s\Xamarin.Forms.Platform.iOS\CollectionView\ObservableItemsSource.cs:165 
  at Xamarin.Forms.Platform.iOS.ObservableItemsSource.CollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs args) [0x00061] in D:\a\1\s\Xamarin.Forms.Platform.iOS\CollectionView\ObservableItemsSource.cs:117 
  at Xamarin.Forms.Platform.iOS.ObservableItemsSource.CollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs args) [0x000b7] in D:\a\1\s\Xamarin.Forms.Platform.iOS\CollectionView\ObservableItemsSource.cs:108 
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1021 
  at Foundation.NSAsyncSynchronizationContextDispatcher.Apply () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/13.20.2.2/src/Xamarin.iOS/Foundation/NSAction.cs:178 
--- End of stack trace from previous location where exception was thrown ---

```

### Basic Information

- Version with issue: 4.8
- Last known good version: 4.5
- IDE: VS Windows
- Platform Target Frameworks:
  - iOS:  latest stable
- Affected Devices: iOS simulator

### Reproduction Link

Repro app: https://github.com/softlion/ReproFormsCollectionViewBug
Run the app on an iOS simulator or device.

Tested on iOS 12.4 / iPhone 6 simulator.
But happens on all iOS OS and devices.

Does not happen on Android / UWP.

This app will trigger one of the 3 above crashes, which comes from different synchronization bugs.
The key is the change of the IsVisible property just before or after items have been added to the collectionviewsource.

### Workaround

None

### Other infos

I've checked the xamarin forms source code.

I'm pretty sure you can fix it easily by making BatchUpdate return a Task, then start BatchUpdate()  with `await _batchUpdating.WaitAsync();` instead of putting `_batchUpdating.Wait()` inside the `PerformBatchUpdates` action parameter.

### LINKED ISSUE #12080: [Bug] CollectionView throws in iOS with a simple example
### Description

I have 2 collection views that are shown by toggling 2 buttons on a form.  When switching between them, iOS throws an exception.  Works fine on Android.

### Steps to Reproduce

1. Run the sample app
2. Click the buttons at the bottom... first Meters, then Tanks, then Meters, exception occurs

image

image

I have seen an exception thrown from here as well...

image

The top level Load() call wraps everything with `Xamarin.Forms.Device.BeginInvokeOnMainThread`.

### Basic Information

- Version with issue: 4.8.0.1364
- IDE: VS 2019

image

- Nuget Packages:

image

- Affected Devices: 
iOS only

### Reproduction Link

Download sample code to reproduce the issue from my [GitHub page](https://github.com/JohnLivermore/CollectionViewIssue).  This code was created today with `File/New Project` (so it's not anything legacy related).

### Workaround

none at this time

### PR DESCRIPTION
**Pull request by hartez**
### Description of Change ###

This is a partial backport of the fixes from PR #12837; some of the changes necessary for that fix also addressed the problems pointed out in #11853. These changes apply those fixes to version 4.8.

### Issues Resolved ### 

- fixes #11853
- fixes #12080

### API Changes ###

None

### Platforms Affected ### 

- iOS

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

Automated UI test

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Review by rmarinho — CHANGES_REQUESTED** [review]
AddingItemToUnviewedCollectionViewShouldNotCrash is failing on iOS11 and iOS12

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]
Captura de pantalla 2020-12-14 a las 14 02 25

File: .Xamarin.Forms.iOS.slnf [hunk 1]
```diff
@@ -23,12 +19,10 @@
       "Xamarin.Forms.Maps\\Xamarin.Forms.Maps.csproj",
       "Xamarin.Forms.Material.iOS\\Xamarin.Forms.Material.iOS.csproj",
       "Xamarin.Forms.Pages.Azure\\Xamarin.Forms.Pages.Azure.csproj",
-      "Xamarin.Forms.Pages.UnitTests\\Xamarin.Forms.Pages.UnitTests.csproj",
       "Xamarin.Forms.Pages\\Xamarin.Forms.Pages.csproj",
+      "Xamarin.Forms.Platform.iOS.UnitTests\\Xamarin.Forms.Platform.iOS.UnitTests.csproj",
```
Comment (jsuarezruiz) [inline .Xamarin.Forms.iOS.slnf:23]: Are these changes necessary?

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue11853.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+				TestCommand = new Command(async () =>
+				{
+					var items = CreateItems(10, 0).ToList().First();
+					var items2 = CreateItems(10, 0).ToList().Skip(1).First();
+
+					int iterations = 1000; //10000;
+
+					for (var i = 0; i < iterations; i++)
+					{
+						await Task.Delay(1);
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue11853.xaml.cs:90]: The example is great, but to cover all cases, could we add to the example?
- Clear all the items and add again.
- Swap the collection.

Maybe add another two buttons for these cases.

**Review by jsuarezruiz — CHANGES_REQUESTED** [review]

File: .Xamarin.Forms.iOS.slnf (hunk 1 above)
Comment (hartez) [reply .Xamarin.Forms.iOS.slnf:23]: Yes, I'm tired of having to add that project back into the solution filter every time I want to build.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue11853.xaml.cs (hunk 2 above)
Comment (hartez) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue11853.xaml.cs:90]: What do you mean by "swap the collection"?

Comment (hartez) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue11853.xaml.cs:90]: > Clear all the items and add again.

That's what it already does - it adds everything, clears it, and adds again in a loop. 1000 times.

Comment (jsuarezruiz) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue11853.xaml.cs:90]: With swap, I mean create a new `ObservableCollection` and assign to Items.

**Review by rmarinho — APPROVED** [review]

**Comment by rmarinho** [comment]
Seems tests are working fine:
https://statics.teams.cdn.office.net/evergreen-assets/safelinks/1/atp-safelinks.html

**Review by jsuarezruiz — APPROVED** [review]

**Comment by jsuarezruiz** [comment]
Could this PR fix https://github.com/xamarin/Xamarin.Forms/issues/10329?

**Comment by rmarinho** [comment]
@hartez seems some tests are failing ios11 and ios12

**Comment by hartez** [comment]
> Could this PR fix #10329?

Tested, and it doesn't.

**Comment by ThumbGen** [comment]
This is a bad fix. Looking forward to seeing others replies. Some of my CollectionViews became unbelievable slow, some don't display any UI elements anymore. 
And no, I won't provide any sample to reproduce, I am personally tired of your CollectionView.
